website/main/content/blog/2025-06-30-sector-expiration.md

```diff
@@ -0,0 +1,213 @@
+  that they can archive the most recent history without having to re-plot, maintaining all the nice
+  properties for sector expiration of the original Subspace protocol.
+
+### Putting all pieces together
+
+With this, we have all that we needed to finally implement farming in shards:
+
+- Plot IDs are unique and stable, allowing us to allocate space into shards.
+- Sectors are committed to a range of history sizes determined by the plot ID, preventing farmers
+  from gaming shard allocation.
+- Piece selection, sector expiration stay unchanged.
+- Farming stays slightly unchanged, apart from the additional information required to verify that
+  the chunk for the piece in the solution range belongs to a sector bound to a plot currently
+  assigned to the shard.
+
+This is how the verification logic for a solution in a shard would look like:
+
+- Standard Subspace Verification (Unchanged): First, perform all existing Subspace checks
+
+```rust
+// Verify piece offset is valid
+if piece_offset >= max_pieces_in_sector {
+    return Err("Invalid piece offset");
+}
+
+// Check sector expiration
+if sector_is_expired(sector_id, current_history_size) {
+    return Err("Sector has expired");
+}
+
+// Verify the piece is part of blockchain history
+if !verify_piece_inclusion(piece, segment_root, proof) {
+    return Err("Invalid piece proof");
+}
+```
+
+- Plot Identity Verification: Verify that the solution comes from a valid plot:
+
+```rust
+// Step 2.1: Compute expected plot ID from farmer's public key and offset
+expected_plot_id = hash(farmer_public_key || plot_offset)
+
+// Step 2.2: Verify the sector was created with this plot ID
+// The sector_id should incorporate the plot_id
+if !sector_belongs_to_plot(sector_id, expected_plot_id) {
+    return Err("Sector doesn't belong to claimed plot");
+}
+```
+
+- Range Validation: Ensure the sector's committed history size is within the plot's valid range:
+
+```rust
+// Step 3.1: Calculate the plot's base range from its offset
+window_level = plot_offset / OFFSETS_PER_WINDOW
+window_size = BASE_WINDOW_SIZE * (2 ^ window_level)
+offset_in_window = plot_offset % OFFSETS_PER_WINDOW
+
+min_history = (window_size * offset_in_window / OFFSETS_PER_WINDOW) - GENESIS_OFFSET
+max_history = (window_size * (offset_in_window + 1) / OFFSETS_PER_WINDOW) - GENESIS_OFFSET
+
+// Step 3.2: Calculate effective range (extends with blockchain growth)
+effective_max_history = max(max_history, current_history_size)
+
+// Step 3.3: Verify sector's history size is within range
+if sector_history_size < min_history || sector_history_size > effective_max_history {
+    return Err("Sector history size outside plot's valid range");
+}
+```
+
+- Shard Assignment Verification: Confirm the plot is assigned to the current shard:
+
+```rust
+// Step 4.1: Determine which shard this plot belongs to
+// Use the plot ID and base range for stable assignment
+expected_shard_id = VRF(plot_id || min_history) % NUM_SHARDS
```

Comment:
`plot_id` is a hash, it can't be used to verify VRF. It is more like we'll use VRF signature to determine shard ID and then verify it against public key of the block seal to see if it is valid.

Reply:
I updated the code, hopefully it is clearer now